nf_core/modules.py

```diff
@@ -231,3 +236,325 @@ def has_valid_pipeline(self):
+                test_nf = test_nf.replace("SUBTOOL", subtool).replace("TOOL", tool)
+                test_nf = test_nf.replace("tool_subtool", tool_name)
+                test_nf = test_nf.replace("TOOL_SUBTOOL", tool_name.upper())
+                test_yml = test_yml.replace("subtool", subtool).replace("tool_", tool + "_")
+                test_yml = test_yml.replace("SUBTOOL", subtool).replace("TOOL", tool)
+                test_yml = re.sub("tool", tool, test_yml)
+            else:
+                meta_yml = meta_yml.replace("tool subtool", tool_name).replace("tool_subtool", "")
+                meta_yml = re.sub("^tool", tool_name, meta_yml)
+                test_nf = (
+                    test_nf.replace("TOOL_SUBTOOL", tool.upper()).replace("SUBTOOL/", "").replace("TOOL", tool.upper())
+                )
+                test_yml = test_yml.replace("tool subtool", tool_name).replace("tool_subtool", "")
+                test_yml = re.sub("^tool", tool_name, test_yml)
+
+            # Install main module files
+            try:
+                os.makedirs(tool_dir, exist_ok=True)
+                # main.nf
+                with open(os.path.join(tool_dir, "main.nf"), "w") as fh:
+                    fh.write(module_nf)
+                # meta.yml
+                with open(os.path.join(tool_dir, "meta.yml"), "w") as fh:
+                    fh.write(meta_yml)
+                # functions.nf
+                with open(os.path.join(tool_dir, "functions.nf"), "w") as fh:
+                    fh.write(functions_nf)
+
+                # Install test files
+                os.makedirs(test_dir, exist_ok=True)
+                # main.nf
+                with open(os.path.join(test_dir, "main.nf"), "w") as fh:
+                    fh.write(test_nf)
+                # test.yml
+                with open(os.path.join(test_dir, "test.yml"), "w") as fh:
+                    fh.write(test_yml)
+            except OSError as e:
+                log.error(f"Could not create module files: {e}")
+                return False
+
+            # Add line to filters.yml
+            try:
+                with open(os.path.join(directory, ".github", "filters.yml"), "a") as fh:
+                    if subtool:
+                        content = (
+                            "\n"
+                            + f"{tool_name}:"
+                            + "\n"
+                            + f"  - software/{tool}/{subtool}/**"
+                            + "\n"
+                            + f"  - tests/software/{tool}/{subtool}/**\n"
+                        )
+                    else:
+                        content = (
+                            "\n"
+                            + f"{tool_name}:"
+                            + "\n"
+                            + f"  - software/{tool}/**"
+                            + "\n"
+                            + f"  - tests/software/{tool}/**\n"
+                        )
+                    fh.write(content)
```

Comment:
```suggestion
                        content = [
                            f"{tool_name}:",
                            f"  - software/{tool}/{subtool}/**",
                            f"  - tests/software/{tool}/{subtool}/**\n",
                        ]
                    else:
                        content = [
                            f"{tool_name}:",
                            f"  - software/{tool}/**",
                            f"  - tests/software/{tool}/**\n",
                        ]
                    fh.write("\n"+"\n".join(content))
```

Note that I'm always a bit scared about manually writing YAML / JSON etc like this. It has a habit of breaking in unexpected ways. Generally parsing and dumping is a bit safer (and often cleaner code).

Reply:
good point .... should rewrite this to parse the `filters.yml` and then dump it properly.